### PR TITLE
DEV-007

### PR DESCRIPTION
**Pull request by Fonthom**

### CODE REVIEW

File: src/main/java/be/syntra/devshop/DevshopBack/controllers/CartController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package be.syntra.devshop.DevshopBack.controllers;
+
+import be.syntra.devshop.DevshopBack.models.CartDto;
+import be.syntra.devshop.DevshopBack.services.CartService;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopBack/controllers/CartController.java]: add @Requestmapping with (temporary) url.

File: src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    @Column(name = "cartCreationDateTime")
+    @JsonFormat
+            (shape = JsonFormat.Shape.STRING, pattern = "dd-MM-yyyy hh:mm:ss")
+
+    private LocalDateTime cartCreationDateTime;
+
+    @NotNull
+    @Column(name = "products")
+    @OneToMany
+    @Cascade(org.hibernate.annotations.CascadeType.ALL)
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: remove @Cascade and add to @OneToMany(cascase = CascadeType.ALL). This way, we don't mix javax.persistence and Spring annotations

File: src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+@NoArgsConstructor
+public class Cart {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "cart_id")
+    private Long id;
+
+
+    @OneToOne(targetEntity = User.class, mappedBy = "activeCart")
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: try to be consistent. activeCart -> active_cart

File: src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    @JsonIgnore
+    private User user;
+
+    @NotNull
+    @Column(name = "cartCreationDateTime")
+    @JsonFormat
+            (shape = JsonFormat.Shape.STRING, pattern = "dd-MM-yyyy hh:mm:ss")
+
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: remove space. No floating fields please

File: src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java (hunk 3 above)
Comment (Paul-Gerarts) [reply src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: please change all the names with camelcasing.

File: src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java (hunk 2 above)
Comment (Paul-Gerarts) [reply src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: please do so for all files with this annotation

File: src/main/java/be/syntra/devshop/DevshopBack/services/utilities/CartMapperUtility.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                .activeCart(cartDto.isActiveCart())
+                .finalizedCart(cartDto.isFinalizedCart())
+                .paidCart(cartDto.isPaidCart())
+                .build();
+    }
+
+    public static List<CartDto> convertToCartDtoList(List<Cart> carts) {
+        return carts.stream()
+                .map(CartMapperUtility::convertToCartDto)
+                .collect(toList());
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopBack/services/utilities/CartMapperUtility.java]: to enhance data-protection, use toUnmodifiableList(). Change this everywhere you're collecting to a list that doesn't need to be modified by us.

File: src/test/java/be/syntra/devshop/DevshopBack/controllers/UserControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        when(userService.findAll()).thenReturn(userList);
+        // when
+        MvcResult response =
+                mockMvc.perform(
+                        get("/users").accept(MediaType.APPLICATION_JSON))
+                        .andExpect(status().isOk())
+                        .andReturn();
+
+
+        // then
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/controllers/UserControllerTest.java]: then: here you will check if the 'reponse' is ok by using:
`response.andExpect(...)`
also: in the this test's given it says to return a userList when service.findAll() is called, the /get will do that so you should check if its in the response here, if there's data send back after the call, you should check the data, pls look at **testRetrieveAllProductsEndpoint**

File: src/test/java/be/syntra/devshop/DevshopBack/services/utilities/AddressMapperUtilityTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+        // then
+        assertEquals(mappedAddress.getClass(), Address.class);
+        assertEquals(mappedAddress.getStreet(), address.getStreet());
+        assertEquals(mappedAddress.getNumber(), address.getNumber());
+        assertEquals(mappedAddress.getBoxNumber(), address.getBoxNumber());
+        assertEquals(mappedAddress.getPostalCode(), address.getPostalCode());
+        assertEquals(mappedAddress.getCity(), address.getCity());
+        assertEquals(mappedAddress.getProvince(), address.getProvince());
+        assertEquals(mappedAddress.getCountry(), address.getCountry());
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/services/utilities/AddressMapperUtilityTest.java]: here you want to check if the result of the mapping (lappedAddress) has the same data as the thing you put in the mapper (addresDto) the first line is checking if the class is correct, thats good, but the rest should be a check between mappedAddress-fields and addresDto-fields, and the `Address address...` line should be removed from // given

File: src/test/java/be/syntra/devshop/DevshopBack/services/utilities/AddressMapperUtilityTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+        // then
+        assertEquals(mappedAddressDto.getClass(), AddressDto.class);
+        assertEquals(mappedAddressDto.getStreet(), addressDto.getStreet());
+        assertEquals(mappedAddressDto.getNumber(), addressDto.getNumber());
+        assertEquals(mappedAddressDto.getBoxNumber(), addressDto.getBoxNumber());
+        assertEquals(mappedAddressDto.getPostalCode(), addressDto.getPostalCode());
+        assertEquals(mappedAddressDto.getCity(), addressDto.getCity());
+        assertEquals(mappedAddressDto.getProvince(), addressDto.getProvince());
+        assertEquals(mappedAddressDto.getCountry(), addressDto.getCountry());
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/services/utilities/AddressMapperUtilityTest.java]: pls compare fields of mappedAddress and address, addressDto has no purpose here, the class-check is ok
you want to check if the mapper functions ok, so you compare it's result (=mappedAddressDto) with what you sent to it (=address).

File: src/test/java/be/syntra/devshop/DevshopBack/services/utilities/ProductMapperUtilityTest.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        Product product = ProductUtils.createProduct();
+        ProductDto productDto = ProductUtils.createProductDto();
+
+        // when
+        Product mappedProduct = ProductMapperUtility.convertToProduct(productDto);
+
+        // then
+        assertEquals(mappedProduct.getClass(), Product.class);
+        assertEquals(mappedProduct.getName(), product.getName());
+        assertEquals(mappedProduct.getPrice(), product.getPrice());
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/services/utilities/ProductMapperUtilityTest.java]: please compare mapper input with mapper output -> mappedProduct <-> productDto

File: src/test/java/be/syntra/devshop/DevshopBack/services/utilities/UserMapperUtilityTest.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        assertEquals(mappedUser.getAddress().getBoxNumber(), user.getAddress().getBoxNumber());
+        assertEquals(mappedUser.getAddress().getPostalCode(), user.getAddress().getPostalCode());
+        assertEquals(mappedUser.getAddress().getCity(), user.getAddress().getCity());
+        assertEquals(mappedUser.getAddress().getProvince(), user.getAddress().getProvince());
+        assertEquals(mappedUser.getAddress().getCountry(), user.getAddress().getCountry());
+        assertEquals(mappedUser.getActiveCart().getProducts(), user.getActiveCart().getProducts());
+        assertEquals(mappedUser.getActiveCart().getCartCreationDateTime(), user.getActiveCart().getCartCreationDateTime());
+        assertEquals(mappedUser.getArchivedCarts().get(0).getCartCreationDateTime(), user.getArchivedCarts().get(0).getCartCreationDateTime());
+        assertEquals(mappedUser.getArchivedCarts().get(0).getProducts().get(0).getName(), user.getArchivedCarts().get(0).getProducts().get(0).getName());
+        assertEquals(mappedUser.getArchivedCarts().get(0).getProducts().get(0).getPrice(), user.getArchivedCarts().get(0).getProducts().get(0).getPrice());
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/services/utilities/UserMapperUtilityTest.java]: pls compare mappers input to mapper output: mappedUser <-> userDto

File: src/test/java/be/syntra/devshop/DevshopBack/services/utilities/UserMapperUtilityTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        assertEquals(mappedUserDto.getAddress().getPostalCode(), userDto.getAddress().getPostalCode());
+        assertEquals(mappedUserDto.getAddress().getCity(), userDto.getAddress().getCity());
+        assertEquals(mappedUserDto.getAddress().getProvince(), userDto.getAddress().getProvince());
+        assertEquals(mappedUserDto.getAddress().getCountry(), userDto.getAddress().getCountry());
+        assertEquals(mappedUserDto.getActiveCart().getCartCreationDateTime(), userDto.getActiveCart().getCartCreationDateTime());
+        assertEquals(mappedUserDto.getActiveCart().getProducts(), userDto.getActiveCart().getProducts());
+        assertEquals(mappedUserDto.getArchivedCarts().get(0).getCartCreationDateTime(), user.getArchivedCarts().get(0).getCartCreationDateTime());
+        assertEquals(mappedUserDto.getArchivedCarts().get(0).getProducts().get(0).getName(), convertToProductDtoList(user.getArchivedCarts().get(0).getProducts()).get(0).getName());
+        assertEquals(mappedUserDto.getArchivedCarts().get(0).getProducts().get(0).getPrice(), convertToProductDtoList(user.getArchivedCarts().get(0).getProducts()).get(0).getPrice());
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/services/utilities/UserMapperUtilityTest.java]: pls compare mapper input with mapper output: mappedUserDto <-> user

File: src/test/java/be/syntra/devshop/DevshopBack/testutilities/JsonUtils.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    public JsonUtils(ObjectMapper mapper) {
+        this.mapper = mapper;
+    }
+
+
+    public String asJsonString(final Object object) {
+        try {
+            return this.mapper.registerModule(new JavaTimeModule())
+                    .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
+                    .writeValueAsString(object);
```
Comment (lenshuygh) [inline src/test/java/be/syntra/devshop/DevshopBack/testutilities/JsonUtils.java]: normally i would expect setting the properties of the mapper outside of this method, but that will only be needed if this class gets a second method that relies on the same settings.

File: src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java (hunk 3 above)
Comment (Fonthom) [reply src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: mappedBy = "activeCart" refers to private Cart activeCart in the User class.
If you would make it active_cart it wil give an error.

error

Comment (Fonthom) [reply src/main/java/be/syntra/devshop/DevshopBack/entities/Cart.java]: ah you placed comment above the line in question

**Pull request merged**